docker/server/entrypoint.sh

```diff
@@ -3,164 +3,213 @@
 
-    for f in /docker-entrypoint-initdb.d/*; do
-        case "$f" in
-            *.sh)
-                if [ -x "$f" ]; then
-                    echo "$0: running $f"
-                    "$f"
-                else
-                    echo "$0: sourcing $f"
-                    # shellcheck source=/dev/null
-                    . "$f"
-                fi
-                ;;
-            *.sql)    echo "$0: running $f"; "${clickhouseclient[@]}" < "$f" ; echo ;;
-            *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${clickhouseclient[@]}"; echo ;;
-            *)        echo "$0: ignoring $f" ;;
-        esac
         echo
-    done
 
-    if ! kill -s TERM "$pid" || ! wait "$pid"; then
-        echo >&2 'Finishing of ClickHouse init process failed.'
-        exit 1
+        # create default database, if defined
+        if [ -n "$CLICKHOUSE_DB" ]; then
+            echo "$0: create database '$CLICKHOUSE_DB'"
+            "${clickhouseclient[@]}" -q "CREATE DATABASE IF NOT EXISTS $CLICKHOUSE_DB";
+        fi
+
+        for f in /docker-entrypoint-initdb.d/*; do
+        echo $f
+            case "$f" in
+                *.sh)
+                    if [ -x "$f" ]; then
+                        echo "$0: running $f"
+                        "$f"
+                    else
+                        echo "$0: sourcing $f"
+                        # shellcheck source=/dev/null
+                        . "$f"
+                    fi
+                    ;;
+                *.sql)    echo "$0: running $f"; "${clickhouseclient[@]}" < "$f" ; echo ;;
+                *.sql.gz) echo "$0: running $f"; gunzip -c "$f" | "${clickhouseclient[@]}"; echo ;;
+                *)        echo "$0: ignoring $f" ;;
+            esac
+            echo
+        done
+
+        if ! kill -s TERM "$pid" || ! wait "$pid"; then
+            echo >&2 'Finishing of ClickHouse init process failed.'
+            exit 1
+        fi
+    fi
+}
+
+_main() {
+    if [ "${1:0:1}" = '-' ]; then
+		set -- clickhouse-server "$@"
+	fi
+    if [ "$1" = 'clickhouse-server' ]; then
+        get_variables
+        # Modify dir permission, create users and parse init_files
+        # only if data directory is empyt
+        if [ -z "$DATABASE_ALREADY_EXISTS" ]; then
+            setup_dirs
+            create_clickhouse_user
+            docker_process_init_files
+        fi
     fi
-fi
 
-# if no args passed to `docker run` or first argument start with `--`, then the user is passing clickhouse-server arguments
-if [[ $# -lt 1 ]] || [[ "$1" == "--"* ]]; then
-    exec $gosu /usr/bin/clickhouse-server --config-file="$CLICKHOUSE_CONFIG" "$@"
-fi
+    # SERVER_ARGS and gosu variables are defined in get_variables 
+    # and set_user functions.
+    # These variables are defined only if the first argument is clickhouse-server
+    exec $gosu "$@" $SERVER_ARGS
```

Comment:
Compatibility: Now it will run custom cmd as 'clickhouse' by default. Before it was root.

Reply:
Even in the original image if i run a custom cmd the user was root:

```
docker run --rm -it yandex/clickhouse-server whoami
root
```

Reply:
Same for the new entrypoint.sh

```
docker run --rm -it garutilorenzo/clickhouse whoami
root
```
gosu variable is empty if the command passed to the entrypoint is different from clickhouse-server